Wiregraph enhancements

MMaiero opened the pull request:
This is a provisional PR over a work in progress.
The branch still needs to be cleaned up, re-enabled the full gwt build and squash the commits.

MMaiero committed:
First drop of new and reviewed APIs for Wires.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Added new ConfigurationService APIs.

Modified import package for related bundles.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Added DriverServiceImpl tests.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Modified gwt server to use the new DriverService apis.

Updated versions and solved missing import in manifest.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

nicolatimeus committed:
Implemented new ConfigurationService apis, added tests

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

MMaiero committed:
Updated apis adding kuraException.

Added code to update wireservice config.
Started implementing WireGraphService.
Updated and added new tests.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Fixed missing import.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Added delete code.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Added code for delete method.

Added corresponding tests.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Remapped the existing WireService impl to new WireGraph impl.

Updated the tests.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Updated tests.

Added javadoc and copyright header.
Removed WireServiceOptions.java

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Modify update method to merge the previous configuration from configService.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

nicolatimeus committed:
First set of Web Ui changes

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

nicolatimeus committed:
Changed handling of driver.pid in asset config

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

MMaiero committed:
Added marshalling APIs.

Introduced JSON encoder/decoder service.
Adapted code to use this service.
Modified distrib and tests accordingly.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Started marshalling/unmarshalling externalization for xml.

Updated code and removed old references.
Updated tests.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

nicolatimeus committed:
Second drop of Web Ui changes

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

MMaiero committed:
Modified wireservice logic.

Reintroduced a tracker with few changes. Simplified the execution path.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Added xml tests grouping from the old ones.

Cleanups.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

MMaiero committed:
Changed visibility in XmlMarshallerImpl.java.

Updated test.
Other minor improvements.

Uniformed names in JSON marshaller/unmarshaller.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

markoer commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/driver/DriverDescriptor.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * The Class DriverDescriptor is responsible for storing the driver instance
+ * configuration. This class can then be passed for serialization and is used to
+ * map the all the information relative to a driver instance. <br>
```
One 'the' too many. Relative => related.

markoer commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/marshalling/Marshalling.java:
```diff
@@ -0,0 +1,50 @@
+ * @since 1.4
+ */
+@ProviderType
+public interface Marshalling {
```
Better name it Marshaller?

MMaiero replied:
I'm still not sure about the name. But Marshaller is maybe a good option

markoer commented on kura/org.eclipse.kura.api/src/main/java/org/eclipse/kura/wire/WireService.java:
```diff
@@ -26,6 +26,7 @@
  * @since 1.2
  */
 @ProviderType
+@Deprecated
```
Add some explanation for this deprecation. It's very frustrating if the replacement is not even mentioned in such cases.

MMaiero replied:
Agreed

markoer commented on kura/org.eclipse.kura.core.configuration/src/main/java/org/eclipse/kura/core/configuration/CloudConfigurationHandler.java:
```diff
@@ -413,27 +415,71 @@ private void doExecSnapshot(CloudletTopic reqTopic, KuraPayload reqPayload, Kura
+        XmlComponentConfigurations result = null;
+        ServiceReference<Marshalling>[] marshallerSRs = getXmlMarshallers();
+        try {
+            for (final ServiceReference<Marshalling> marshallerSR : marshallerSRs) {
```
Maybe break on success?

MMaiero replied:
I thought that too, but Sonar was complaining. I will review

markoer reviewed (commented):
First part...

nicolatimeus committed:
Updated Drivers and Assets section

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

markoer commented on kura/org.eclipse.kura.core.configuration/src/main/java/org/eclipse/kura/core/configuration/upgrade/ConfigurationUpgrade.java:
```diff
@@ -63,11 +89,145 @@ public static XmlComponentConfigurations upgrade(XmlComponentConfigurations xmlC
+                    continue;
+                }
+                if (key.startsWith(String.format(PATTERN, wireConfId))) {
+                    if (key.contains("emitter")) {
```
This might present problems with localization, I believe. If a language is added that doesn't use 'emitter' and 'receiver'.

markoer commented on kura/org.eclipse.kura.core.configuration/src/main/java/org/eclipse/kura/core/configuration/util/ComponentUtil.java:
```diff
@@ -268,11 +273,11 @@ public static Tocd readObjectClassDefinition(String pid)
         if (metatypeXml != null) {
-            Tmetadata metaData = XmlUtil.unmarshal(metatypeXml, Tmetadata.class);
+            Tmetadata metaData = unmarshalXml(metatypeXml, Tmetadata.class);
             if (metaData.getOCD() != null && metaData.getOCD().size() > 0) {
```
!isEmpty()?

markoer commented on kura/org.eclipse.kura.core.configuration/src/main/java/org/eclipse/kura/core/configuration/util/ComponentUtil.java:
```diff
@@ -302,7 +307,7 @@ public static Tocd readObjectClassDefinition(Bundle bundle, String pid)
         if (metatypeXml != null) {
-            Tmetadata metaData = XmlUtil.unmarshal(metatypeXml, Tmetadata.class);
+            Tmetadata metaData = unmarshalXml(metatypeXml, Tmetadata.class);
             if (metaData.getOCD() != null && metaData.getOCD().size() > 0) {
```
!isEmpty()?

nicolatimeus committed:
Fixed handling of Asset non channel related configuration fields

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

markoer commented on kura/org.eclipse.kura.core.deployment/src/main/java/org/eclipse/kura/core/deployment/CloudDeploymentHandlerV2.java:
```diff
@@ -955,4 +957,29 @@ public void installDownloadedFile(File dpFile, DeploymentPackageInstallOptions o
     }
 
+    private ServiceReference<Marshalling>[] getXmlMarshallers() {
+        String filterString = String.format("(&(kura.service.pid=%s))", "org.eclipse.kura.marshalling.xml.provider");
```
Call ComponentUtil.getXmlMarshallers() here and similar in the following 2 methods.

MMaiero committed:
First set of changes related to review.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

markoer reviewed (commented):
Second part

markoer commented on kura/test/org.eclipse.kura.core.configuration.test/src/test/java/org/eclipse/kura/core/configuration/ConfigurationServiceTest.java:
```diff
@@ -3003,4 +3095,165 @@ String getSnapshotsDirectory() {
+            List<Tocd> registeredOcds, List<Component> registeredComponents) throws NoSuchFieldException {
+
+        assertEquals(registeredFactories.size(), registeredOcds.size());
+        ScrService scrService = new ScrService() {
```
Since most methods return null, it could be a mock that just reacts to getComponents().

nicolatimeus replied:
Yes definitely, I've tried using a mock here but I was getting some strange "unfinished stubbing" errors at runtime. I did not spend much time investigating and I've used this quick and dirty solution. I will try again and see if I can manage to use a proper mock.

markoer reviewed (commented):
One more about the tests. The rest will follow next week.

MMaiero committed:
Removed WireService independent thread.

Changed the tracking.
Minor cleanups.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

markoer commented on kura/test/org.eclipse.kura.core.test/src/main/java/org/eclipse/kura/core/test/AllCoreTests.java:
```diff
@@ -33,7 +33,7 @@
 @SuiteClasses({ DataServiceTest.class, CloudDeploymentHandlerTest.class, CloudServiceTest.class, CommURITest.class,
         ComponentConfigurationImplTest.class, ConfigurationServiceTest.class, NetUtilTest.class,
-        NetworkServiceTest.class, SystemAdminServiceTest.class, XmlUtilTest.class })
+        NetworkServiceTest.class, SystemAdminServiceTest.class })
```
Rename the static fields, since the class is modified, anyway.

markoer commented on kura/test/org.eclipse.kura.wire.provider.test/src/main/java/org/eclipse/kura/wire/provider/test/WireServiceTest.java:
```diff
@@ -87,6 +83,7 @@ public void bindWireService(final WireService wireService) {
      */
     @TestTarget(targetPlatforms = { TestTarget.PLATFORM_ALL })
     @Test
+    @Ignore
```
Are these no longer applicable or just stopped working?

markoer commented on kura/test/org.eclipse.kura.xml.marshaller.unmarshaller.provider.test/src/test/java/org/eclipse/kura/internal/xml/marshaller/unmarshaller/test/XmlEncoderDecoderTest.java:
```diff
@@ -26,57 +24,38 @@
+            String marshalledString = xmlMarshallerImpl.marshal(xmlComponentConfigurations);
+            Object unmarshalledObjectFromString = xmlMarshallerImpl.unmarshal(marshalledString,
+                    XmlComponentConfigurations.class);
             Assert.assertTrue(
```
import static ...Assert.assertTrue

markoer commented on kura/test/org.eclipse.kura.xml.marshaller.unmarshaller.provider.test/src/test/java/org/eclipse/kura/internal/xml/marshaller/unmarshaller/test/XmlEncoderDecoderTest.java:
```diff
@@ -367,25 +345,77 @@ private static XmlComponentConfigurations getSampleXmlComponentConfigurationsObj
+        xcc.setConfigurations(ccis);
+
+        String s = xmlMarshallerImpl.marshal(xcc);
+        System.err.println(s);
```
Use logger instead, if necessary.

markoer commented on kura/test/org.eclipse.kura.xml.marshaller.unmarshaller.provider.test/src/test/java/org/eclipse/kura/internal/xml/marshaller/unmarshaller/test/XmlEncoderDecoderTest.java:
```diff
@@ -367,25 +345,77 @@ private static XmlComponentConfigurations getSampleXmlComponentConfigurationsObj
+
+        XmlComponentConfigurations config1 = xmlMarshallerImpl.unmarshal(s, XmlComponentConfigurations.class);
+        String s1 = xmlMarshallerImpl.marshal(config1);
+        System.err.println(s1);
```
Use logger instead.

markoer commented on kura/test/org.eclipse.kura.xml.marshaller.unmarshaller.provider.test/src/test/java/org/eclipse/kura/internal/xml/marshaller/unmarshaller/test/XmlEncoderDecoderTest.java:
```diff
@@ -367,25 +345,77 @@ private static XmlComponentConfigurations getSampleXmlComponentConfigurationsObj
+        xcc.setConfigurations(ccis);
+
+        String s = xmlMarshallerImpl.marshal(xcc);
+        System.err.println(s);
```
Use logger instead.

markoer commented on kura/test/org.eclipse.kura.xml.marshaller.unmarshaller.provider.test/src/test/java/org/eclipse/kura/internal/xml/marshaller/unmarshaller/test/XmlEncoderDecoderTest.java:
```diff
@@ -367,25 +345,77 @@ private static XmlComponentConfigurations getSampleXmlComponentConfigurationsObj
+
+        XmlComponentConfigurations config1 = xmlMarshallerImpl.unmarshal(s, XmlComponentConfigurations.class);
+        String s1 = xmlMarshallerImpl.marshal(config1);
+        System.err.println(s1);
```
Use logger instead.

MMaiero committed:
Integrated few review comments.

Solved exception in drivers and assets page.
Tests cleanup.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

markoer commented:
Tables don't work in IE on this branch:
![image](https://user-images.githubusercontent.com/22816717/33554611-b803a750-d8fd-11e7-8e86-1ab505960627.png)

Where rows should be selectable, they are not:
![image](https://user-images.githubusercontent.com/22816717/33554625-cd9d2fc8-d8fd-11e7-8a45-642bddfc9493.png)
![image](https://user-images.githubusercontent.com/22816717/33554676-fec30fa0-d8fd-11e7-9f4f-bf3bfbf16103.png)

List of Wires components is empty, too:
![image](https://user-images.githubusercontent.com/22816717/33554654-e9081840-d8fd-11e7-8283-8c11125c100f.png)

Kura 3.2.0-SNAPSHOT on this branch on RPi3. Client: IE 11.0.9600, document mode IE 9. OK in IE 10 and Edge modes.

markoer commented:
Buttons for driver and asset management are now floating to the right-hand side instead of the left:
![image](https://user-images.githubusercontent.com/22816717/33599501-a2eed106-d9a6-11e7-907a-beea7ae1de91.png)

Update: As @MMaiero stated, it's already fixed.

MMaiero commented:
This last is already solved with yesterday's commit

nicolatimeus commented:
IE11 issues are not yet addressed in last commit

nicolatimeus committed:
Minor Web ui fixes and cleanup

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

nicolatimeus committed:
IE11 fixes

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

nicolatimeus commented:
Last commit should hopefully fix the IE11 issues.
I wasn't able to reproduce the table rendering issues or the exceptions on the log reported in #1821.
I have noticed another issue with IE11 (11.0.9600.17420 running inside a Win 7 vm): dragging and dropping from Eclipse Marketplace between two IE11 windows does not work (IE opens the dropped url in the current window), but it works when dragging from a Chrome window to an IE11 window.

markoer commented:
@nicolatimeus: I can confirm the issue with D&D of package install link from Marketplace.
I cannot reproduce the table issues anymore, so probably IE just decided to work.

nicolatimeus commented:
@markoer it is possible that you tested IE some commits ago, when only the Chrome profile was enabled to speed up the build, so IE-specific GWT code was probably disabled, .
Matteo re-enabled the full build in one of his last commits so probably that fixed the issue with the tables.

nicolatimeus committed:
Used proper mock in configuration test

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

MMaiero committed:
Changes to have driver apis unmodified.

Moved all the needed changes to another package.
Modified the code accordingly.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>

nicolatimeus committed:
Versioning fixes

Signed-off-by: nicolatimeus <nicola.timeus@eurotech.com>

MMaiero committed:
Added OCDService interface.
Moved code to this new service.

Updated tests.
Fixed imports.

Signed-off-by: Maiero <matteo.maiero@eurotech.com>
